File: perf/benchmarks/micro/math_loop/math_loop.cpp
// Math-intensive loop — std math micro-benchmark (C++).
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 5'000'000LL;

    auto t0 = std::chrono::steady_clock::now();
    double s = 0.0;
    double inv = 1.0 / static_cast<double>(n);
    for (int64_t i = 0; i < n; ++i) {
        double x = static_cast<double>(i) * inv;
        s += std::sin(x) * std::cos(x) + std::sqrt(x + 1.0) - std::exp(-x);
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    int64_t cs = static_cast<int64_t>(std::llround(s * 1e6));
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/micro/math_loop/math_loop.py
# Math-intensive loop — std.math micro-benchmark.
#
# Computes a sum of sin/cos/sqrt/exp values over N iterations.
# Stresses the math standard library and floating-point throughput.
#
# Argument: N (default 5_000_000).

import math
import sys
import time


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 5_000_000

    t0 = time.perf_counter_ns()
    s = 0.0
    inv = 1.0 / n
    for i in range(n):
        x = i * inv
        s += math.sin(x) * math.cos(x) + math.sqrt(x + 1.0) - math.exp(-x)
    elapsed_ns = time.perf_counter_ns() - t0

    # Round to integer scale to keep the checksum stable across implementations
    # despite tiny floating-point reordering effects.
    cs = round(s * 1e6)
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
